### PR TITLE
Fix writing of track metadata on Windows  

### PR DESCRIPTION
**Pull request by daschuer**
This fixes https://bugs.launchpad.net/mixxx/+bug/1955314 and https://bugs.launchpad.net/mixxx/+bug/1955331. 

This is a mayor bugfix that should be released soon. Because the track metadata was not exported on windows when the track has not been loaded into a deck. Since the user assumes that the metadata is written, there is a risk of data loss. This was discovered in the first attempt for a fix here: https://github.com/mixxxdj/mixxx/pull/4572 

The fix makes use of the ReplaceFileW() API function that is recommended instead of the deprecated transition feature for atomic file operations. ReplaceFileW() avoids virus scanner and indexer locking and keeps the file creation time at its old value. 

The refactoring part has been issued here https://github.com/mixxxdj/mixxx/pull/4584 for the main branch.

### CODE REVIEW

File: src/sources/metadatasourcetaglib.cpp [hunk 1]
```diff
@@ -702,6 +714,97 @@ class SafelyWritableFile final {
...
+                    return false;
+                case ERROR_SHARING_VIOLATION:
+                    // The process cannot access the file because it is being used by another process.
+                    kLogger.critical()
+                            << "Unable to replace"
+                            << m_origFileName
+                            << "by"
+                            << m_tempFileName
+                            << "because it is used by another process";
+                    Sleep(kWindowsSharingViolationsSleepMs); // Note: Capital S for Windows API Sleep()
```
Comment (uklotzde) [inline src/sources/metadatasourcetaglib.cpp:766]: Why don't you use `QThread::msleep()` here?

File: src/sources/metadatasourcetaglib.cpp [hunk 2]
```diff
@@ -30,6 +34,14 @@ const QString kSafelyWritableTempFileSuffix = QStringLiteral("_temp");
 // potential failures caused by exceeded path length.
 const QString kSafelyWritableOrigFileSuffix = QStringLiteral("_orig");
 
+#ifdef __WINDOWS__
+// After Mixxx has closed the track file, the indexer or virus scanner
+// might kick in and fail ReplaceFileW() with a sharing violation when
+// replacing the original file with the one with the updated metadata.
+const int kWindowsSharingViolationsRetries = 5;
```
Comment (uklotzde) [inline src/sources/metadatasourcetaglib.cpp:41]: kWindowsSharingViolationMaxRetries

File: src/sources/metadatasourcetaglib.cpp [hunk 3]
```diff
@@ -30,6 +34,14 @@ const QString kSafelyWritableTempFileSuffix = QStringLiteral("_temp");
...
+const int kWindowsSharingViolationsSleepMs = 100;
```
Comment (uklotzde) [inline src/sources/metadatasourcetaglib.cpp:42]: kWindowsSharingViolationSleepBeforeNextRetryMillis

File: src/sources/metadatasourcetaglib.cpp [hunk 4]
```diff
@@ -702,6 +714,97 @@ class SafelyWritableFile final {
...
+                    // retain their original file names.
+                    kLogger.critical()
+                            << "Unable to remove"
+                            << m_origFileName
+                            << "before replacing by"
+                            << m_tempFileName;
+                    return false;
+                case ERROR_SHARING_VIOLATION:
+                    // The process cannot access the file because it is being used by another process.
+                    kLogger.critical()
```
Comment (uklotzde) [inline src/sources/metadatasourcetaglib.cpp:760]: Not critical if a retry is following.

File: src/sources/metadatasourcetaglib.cpp [hunk 5]
```diff
@@ -702,6 +714,97 @@ class SafelyWritableFile final {
...
+                            << "Error"
+                            << error
+                            << "during replacing"
+                            << m_origFileName
+                            << "by"
+                            << m_tempFileName;
+                    return false;
+                }
+            }
+            break;
```
Comment (uklotzde) [inline src/sources/metadatasourcetaglib.cpp:791]: It is difficult to recognize that this break belongs to the loop. Is it actually needed?

File: src/sources/metadatasourcetaglib.cpp [hunk 6]
```diff
@@ -30,6 +34,14 @@ const QString kSafelyWritableTempFileSuffix = QStringLiteral("_temp");
 // potential failures caused by exceeded path length.
 const QString kSafelyWritableOrigFileSuffix = QStringLiteral("_orig");
 
+#ifdef __WINDOWS__
+// After Mixxx has closed the track file, the indexer or virus scanner
```
Comment (uklotzde) [inline src/sources/metadatasourcetaglib.cpp:38]: This comment belongs to the code where these constants are actually used, but not to these constants. Mentioning why ReplaceFileW() is needed here but not in the context were it is actually invoked is confusing.

File: src/sources/soundsourceproxy.cpp [hunk 7]
```diff
@@ -361,6 +361,12 @@ SoundSourceProxy::exportTrackMetadataBeforeSaving(Track* pTrack, UserSettingsPoi
                 pTrack->updateStreamInfoFromSource(
                         pAudioSource->getStreamInfo());
                 DEBUG_ASSERT(pTrack->hasStreamInfoFromSource());
+                // Make sure that the track file is closes after reading
```
Comment (uklotzde) [inline src/sources/soundsourceproxy.cpp:364]: multiple typos

File: src/sources/soundsourceproxy.cpp [hunk 8]
```diff
@@ -731,6 +737,7 @@ mixxx::AudioSourcePointer SoundSourceProxy::openAudioSource(
                 // before exporting metadata. In this case the caller (this class)
                 // is responsible for updating the stream info if needed.
                 if (!m_pTrack) {
+                    // TODO() It is surprising to receive here a m_pSoundSource;
```
Comment (uklotzde) [inline src/sources/soundsourceproxy.cpp:740]: - TODO is not a function
- "receive"?
- semicolon to end a comment?

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
Done

File: src/sources/soundsourceproxy.cpp [hunk 9]
```diff
@@ -731,6 +737,7 @@ mixxx::AudioSourcePointer SoundSourceProxy::openAudioSource(
                 // before exporting metadata. In this case the caller (this class)
                 // is responsible for updating the stream info if needed.
                 if (!m_pTrack) {
+                    // TODO: It is surprising that openAudioSource returns a m_pSoundSource
```
Comment (uklotzde) [inline src/sources/soundsourceproxy.cpp:740]: Why is it suprising?? `SoundSource` is an extension of both `AudioSource` and `MetadataSource`, basic OO design.

Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:740]: This should work and fix the bug with the open file:

```
m_pAudioSource = m_pSoundSource;
return m_pAudioSource;
```

File: src/sources/soundsourceproxy.cpp [hunk 10]
```diff
@@ -361,6 +361,12 @@ SoundSourceProxy::exportTrackMetadataBeforeSaving(Track* pTrack, UserSettingsPoi
                 pTrack->updateStreamInfoFromSource(
                         pAudioSource->getStreamInfo());
                 DEBUG_ASSERT(pTrack->hasStreamInfoFromSource());
+                // Make sure that the track file is closed after reading.
```
Comment (uklotzde) [inline src/sources/soundsourceproxy.cpp:364]: This additional code should become obsolete with the fix I just proposed.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/sources/soundsourceproxy.cpp (hunk 10 above)
Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:364]: This is still needed:
```
proxy.closeAudioSource();
```

File: src/sources/soundsourceproxy.cpp (hunk 9 above)
Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:740]: I have already fixed the underlying issue in https://github.com/mixxxdj/mixxx/pull/4584 please have a look.

File: src/sources/soundsourceproxy.cpp (hunk 10 above)
Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:364]: I would like to keep this as an hot fix solution with a minimum impact.

The rest is fixed in https://github.com/mixxxdj/mixxx/pull/4584

Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:364]: I would prefer a proper fix here.

Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:364]: I am not going to merge a hack with a wrong comment.

Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:364]: Why do you think the comment is wrong? Add least it is working.

But I don't mind to merge your ideas.
Can you issue a PR against my branch? 

Thank you.

Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:364]: The alternative would be to merge https://github.com/mixxxdj/mixxx/pull/4584 to this branch. This is IMHO the proper fix, but touches more code.

Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:364]: I haven't reviewed #4584 yet and I cannot promise to do it soon.

Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:364]: https://github.com/daschuer/mixxx/pull/72

Moreover the build failed on Linux with Clang due to a wrong preprocessor directive. Fixed.

**Comment by uklotzde** [comment]
This important bugfix should be mentioned in the CHANGELOG.

**Comment by daschuer** [comment]
We have got the confirmation of that it is working: 
https://bugs.launchpad.net/mixxx/+bug/1955314

**Review by uklotzde — APPROVED** [review]
Thank you! LGTM

I'll merge to main

**Pull request merged**